doc: document ThreadSafeFunction

KevinEady opened the pull request:
Re: https://github.com/nodejs/node-addon-api/pull/442

gabrielschulhof commented on doc/threadsafe_function.md:
````diff
@@ -0,0 +1,219 @@
+Creates a new instance of the `Napi::ThreadSafeFunction` object. The `New` function has several overloads for the various optional parameters: skip the optional parameter for that specific overload.
+
+```cpp
+New(napi_env env,
````
We should change this to `Napi::Env env` - [commented in the PR](https://github.com/nodejs/node-addon-api/pull/442/files#r293508141) as well.

KevinEady replied:
As discussed in the PR, we decided to stick with `napi_env`. Therefore, this was not changed.

gabrielschulhof commented on doc/threadsafe_function.md:
````diff
@@ -0,0 +1,219 @@
+    Context* context);
+```
+
+- `env`: The `napi_env` environment in which to construct the `Napi::ThreadSafeFunction` object.
````
Ditto wrt. changing to `Napi::Env`.

gabrielschulhof commented on doc/threadsafe_function.md:
````diff
@@ -0,0 +1,219 @@
+bool Napi::ThreadSafeFunction::Acquire()
+```
+
+Returns `true` if the thread can successfully be added to the thread count.
````
Maybe change this to `... if the thread has successfully acquired the thread-safe function for its use.` The implication that a successful addition to the thread count implies that the thread may henceforth use this thread-safe function is not necessarily obvious.

KevinEady replied:
The signature for `Acquire()`, `Release()`, `Abort()`, and `BlockingCall()/NonBlockingCall()`  was changed to return an `napi_status`. This documentation has been updated for those return values.

gabrielschulhof commented on doc/threadsafe_function.md:
````diff
@@ -0,0 +1,219 @@
+bool Napi::ThreadSafeFunction::Release()
+```
+
+Returns `true` if the thread-safe function has successfully released.
````
"... has been successfully released.",  maybe?

KevinEady replied:
Addressed; likewise with the signature change mentioned above.

gabrielschulhof commented on doc/threadsafe_function.md:
```diff
@@ -0,0 +1,219 @@
+
+### Abort
+
+"Abort" the thread-safe function. This will cause all subsequent APIs associated with the thread-safe function except `Release()` to return `napi_closing` even before its reference count reaches zero. In particular, `BlockingCall` and `NonBlockingCall()` will return `napi_closing`, thus informing the threads that it is no longer possible to make asynchronous calls to the thread-safe function. Thi ...
```
"&#96;napi_closing&#96;" should be "&#96;CLOSING&#96;".

KevinEady replied:
`napi_closing` is correct as `CLOSING` was the previous `ThreadSafeFunction::Status` enum, which is no longer used.

gabrielschulhof commented on doc/threadsafe_function.md:
```diff
@@ -0,0 +1,219 @@
+
+### BlockingCall / NonBlockingCall
+
+Calls the Javascript function in a either a blocking or non-blocking fashion.
```
"... in either a blocking ... "

KevinEady replied:
Addressed.

gabrielschulhof commented on doc/threadsafe_function.md:
```diff
@@ -0,0 +1,219 @@
+
+Calls the Javascript function in a either a blocking or non-blocking fashion.
+- `BlockingCall()`: the API blocks until space becomes available in the queue. Will never block if the thread-safe function was created with a maximum queue size of `0`.
+- `NonBlockingCall()`: will return `napi_queue_full` if the queue was full, preventing data from being successfully added to the queue
```
Period at the end of the sentence.

KevinEady replied:
Addressed.

KevinEady commented on doc/threadsafe_function.md:
```diff
@@ -0,0 +1,219 @@
+# ThreadSafeFunction
+
+JavaScript functions can normally only be called from a native addon's main thread. If an addon creates additional threads, then node-addon-api functions that require a `Env`, `Value`, or `Ref` must not be called from those threads.
```
Added namespace of `Napi`, changed `Ref` to `Reference`

KevinEady replied:
Lines have been wrapped to 80 characters.

KevinEady commented on doc/threadsafe_function.md:
```diff
@@ -0,0 +1,219 @@
+
+`ThreadSafeFunction` objects are destroyed when every thread which uses the object has called `Release()` or has received a return status of `CLOSING` in response to a call to `BlockingCall()` or `NonBlockingCall()`. The queue is emptied before the `ThreadSafeFunction` is destroyed. It is important that `Release()` be the last API call made in conjunction with a given `ThreadSafeFunction`, becaus ...
+
+Once the number of threads making use of a `ThreadSafeFunction` reaches zero, no further threads can start making use of it by calling `Acquire()`. In fact, all subsequent API calls associated with it, except `Release()`, will return an error value of napi_closing.
```
Removed all old references of the old implementation's `ThreadSafeFunction::Status` to the various `napi_status`es

KevinEady commented:
@gabrielschulhof / @romandev ,

Most recent commit addresses the issues brought up, with signatures now matching the approved #442  PR. Please review.

KevinEady commented on doc/threadsafe_function.md:
```diff
@@ -0,0 +1,303 @@
+the object has called `Release()` or has received a return status of
+`napi_closing` in response to a call to `BlockingCall()` or `NonBlockingCall()`.
+The queue is emptied before the `Napi::ThreadSafeFunction` is destroyed. It is
+important that `Release()` be the last API call made in conjunction with a given
```
One thing (even with the current wording, generally taken from [n_api_asynchronous_thread_safe_function_calls](https://nodejs.org/api/n-api.html#n_api_asynchronous_thread_safe_function_calls)): I don't think the text is correct, because `Release()` [or `Abort()`] don't _necessarily_ have to be the last call. 

In a normal, non-aborted flow, `Release()` would be the last call, but in an aborted flow, it is possible that a `[Non]BlockingCall()` could be the last call, as it also decreases the thread count:

>  This can be used as a criterion for terminating the thread.

Should we better specify the abortable behavior here as well, or is it fine to keep in the Abort section?

gabrielschulhof replied:
`Release()` or `Abort()` *absolutely* have to be the last call – on the thread making the call. I think the most important thing to point out, and the aspect to emphasize here is this: After you've called `Release()` or `Abort()`, make no further use of the TSFN *on that thread*, irrespective of how this will affect the other threads – namely that, in the case of `Abort()` it will cause them to receive `napi_closing`s.

This latter point is documented further down in the paragraph, and `BlockingCall()` and `NonBlockingCall()` are documented to have this feature such that the two calls are presented side-by-side as equal and alternative ways of being informed of the TSFN's impending closure.

The fact is that the documentation in core is almost certainly not phrased ideally. This is actually an opportunity to improve both places of documentation.

mhdawson commented:
Just noticed that I think this might need a link on the main page: https://github.com/nodejs/node-addon-api
